compiler/src/diagnostics/comments.re

```diff
@@ -0,0 +1,112 @@
+open Grain_parsing;
+open Grain_typed;
+
+type comments = list(Typedtree.comment);
+
+let comment_directly_before_location =
+    (loc: Grain_parsing.Location.t, comments: list(Typedtree.comment)) => {
+  let (_file, line, _char) = Location.get_pos_info(loc.loc_start);
+  // Tracks the line exactly before
+  let directlyPrev = line - 1;
+  // Hack to handle code that has an attribute on the line before, such as `@disableGC`
+  let attributePrev = line - 2;
+  List.find_map(
+    (comment: Grain_typed.Typedtree.comment) => {
+      switch (comment) {
+      | Line({cmt_loc})
+      | Shebang({cmt_loc})
+      | Block({cmt_loc})
+      | Doc({cmt_loc}) when cmt_loc.loc_end.pos_lnum == directlyPrev =>
+        Some(comment)
+      | Line({cmt_loc})
+      | Shebang({cmt_loc})
+      | Block({cmt_loc})
+      | Doc({cmt_loc}) when cmt_loc.loc_end.pos_lnum == attributePrev =>
+        Some(comment)
+      // TODO: Handle comment on same line but before the location
+      | _ => None
+      }
+    },
+    comments,
+  );
+};
+
+let docblock_for_value_description =
+    (~comments: comments, vd: Types.value_description) => {
+  let comment = comment_directly_before_location(vd.val_loc, comments);
+  switch (comment) {
+  | Some(Doc({cmt_content})) => Some(cmt_content)
+  | _ => None
+  };
+};
+
+let docblock_for_type_declaration =
+    (~comments: comments, td: Types.type_declaration) => {
+  let comment = comment_directly_before_location(td.type_loc, comments);
+  switch (comment) {
+  | Some(Doc({cmt_content})) => Some(cmt_content)
+  | _ => None
+  };
+};
+
+// Attributes in a comment are prefixed with `@` symbol, such as `@param`
+module Attributes = {
+  type attr_name = string;
+  type attr_type = string;
+  type attr_desc = string;
+
+  type t =
+    | Param(attr_name, attr_type, attr_desc)
+    | Returns(attr_type, attr_desc);
+
+  let extract =
+      (
+        ~args: option(list(Types.type_expr))=?,
+        ~returns: option(Types.type_expr)=?,
+        comment,
+      ) => {
+    let attrs = ref([]);
+    // https://regexper.com/#%5E%40%28param%7Creturns%29%28%5B%20%5Ct%5D%28%5B%5E%3A%5D%2B%29%3A%29%3F%5B%20%5Ct%5D%28.*%29%24
+    let re =
+      Str.regexp(
+        "^@\\(param\\|returns\\)\\([ \t]\\([^:]+\\):\\)?[ \t]\\(.*\\)$",
+      );
```

Comment:
@ospencer I tried to bust out Ocamllex but failed spectacularly. Maybe we can pair on that if you think it is better than this wild regexp.